File: adventofcode/2023/day16/day16.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

type Pos struct{ x, y int }

func (p Pos) up() Pos {
	return Pos{p.x, p.y - 1}
}
func (p Pos) down() Pos {
	return Pos{p.x, p.y + 1}
}
func (p Pos) left() Pos {
	return Pos{p.x - 1, p.y}
}
func (p Pos) right() Pos {
	return Pos{p.x + 1, p.y}
}

func dfs(grid []string, pos Pos, dirn string, vis map[Pos]string) int {
	if pos.x < 0 || pos.y < 0 || pos.x >= len(grid[0]) || pos.y >= len(grid) {
		return len(vis)
	}

	c := grid[pos.y][pos.x]
	if d := vis[pos]; d == dirn {
		return len(vis)
	}

	vis[pos] = dirn
	switch c {
	case '.':
		switch dirn {
		case "up":
			dfs(grid, pos.up(), dirn, vis)
		case "down":
			dfs(grid, pos.down(), dirn, vis)
		case "left":
			dfs(grid, pos.left(), dirn, vis)
		case "right":
			dfs(grid, pos.right(), dirn, vis)
		}
	case '|':
		switch dirn {
		case "up":
			dfs(grid, pos.up(), dirn, vis)
		case "down":
			dfs(grid, pos.down(), dirn, vis)
		case "left", "right":
			dfs(grid, pos.up(), "up", vis)
			dfs(grid, pos.down(), "down", vis)
		}
	case '-':
		switch dirn {
		case "up", "down":
			dfs(grid, pos.left(), "left", vis)
			dfs(grid, pos.right(), "right", vis)
		case "left":
			dfs(grid, pos.left(), dirn, vis)
		case "right":
			dfs(grid, pos.right(), dirn, vis)
		}
	case '/':
		switch dirn {
		case "up":
			dfs(grid, pos.right(), "right", vis)
		case "down":
			dfs(grid, pos.left(), "left", vis)
		case "left":
			dfs(grid, pos.down(), "down", vis)
		case "right":
			dfs(grid, pos.up(), "up", vis)
		}
	case '\\':
		switch dirn {
		case "up":
			dfs(grid, pos.left(), "left", vis)
		case "down":
			dfs(grid, pos.right(), "right", vis)
		case "left":
			dfs(grid, pos.up(), "up", vis)
		case "right":
			dfs(grid, pos.down(), "down", vis)
		}
	}
	return len(vis)
}

func part1(lines []string) int {
	return dfs(lines, Pos{0, 0}, "right", map[Pos]string{})
}

func part2(lines []string) int {
	ymax, xmax := len(lines), len(lines[0])
	maxcount := 0
	for x := 0; x < xmax; x++ {
		c := dfs(lines, Pos{x, 0}, "down", map[Pos]string{})
		if maxcount < c {
			maxcount = c
		}
		c = dfs(lines, Pos{x, ymax - 1}, "up", map[Pos]string{})
		if maxcount < c {
			maxcount = c
		}
	}
	for y := 0; y < ymax; y++ {
		c := dfs(lines, Pos{0, y}, "right", map[Pos]string{})
		if maxcount < c {
			maxcount = c
		}
		c = dfs(lines, Pos{xmax - 1, y}, "left", map[Pos]string{})
		if maxcount < c {
			maxcount = c
		}
	}
	return maxcount
}

func main() {
	s := time.Now()
	lines := readInput()
	s1 := time.Now()
	fmt.Println(part1(lines))
	e1 := time.Since(s1)
	s2 := time.Now()
	fmt.Println(part2(lines))
	e2 := time.Since(s2)
	e := time.Since(s)
	fmt.Printf("Time for part1: %s, part2: %s, total: %s\n", e1, e2, e)
}

File: adventofcode/2023/day16/day16.zig
const std = @import("std");
const time = std.time;
const print = std.debug.print;
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;
const HashMap = std.AutoHashMap;

const Pos = struct {
    x: i32,
    y: i32,

    const Self = @This();

    fn init(x: i32, y: i32) Self {
        return Self{ .x = x, .y = y };
    }

    fn up(self: Self) Self {
        return Self{ .x = self.x, .y = self.y - 1 };
    }
    fn down(self: Self) Self {
        return Self{ .x = self.x, .y = self.y + 1 };
    }
    fn left(self: Self) Self {
        return Self{ .x = self.x - 1, .y = self.y };
    }
    fn right(self: Self) Self {
        return Self{ .x = self.x + 1, .y = self.y };
    }
};

const Direction = enum {
    up,
    down,
    left,
    right,
};

const Cell = enum {
    space,

    leftMirror,
    rightMirror,

    verticalSplitter,
    horizontalSplitter,
};

const Grid = struct {
    grid: ArrayList(ArrayList(Cell)),
    xmax: usize,
    ymax: usize,
    allocator: Allocator,

    const Self = @This();

    fn init(allocator: Allocator) Self {
        const grid = ArrayList(ArrayList(Cell)).init(allocator);
        return Self{ .grid = grid, .xmax = 0, .ymax = 0, .allocator = allocator };
    }

    fn len(self: Self) usize {
        return self.grid.items.len;
    }

    fn append(self: *Self, row: ArrayList(Cell)) !void {
        self.ymax += 1;
        self.xmax = row.items.len;
        return try self.grid.append(row);
    }

    fn deinit(self: Self) void {
        self.grid.deinit();
    }

    fn outside(self: Self, pos: Pos) bool {
        if (pos.x < 0 or pos.y < 0 or pos.x >= self.xmax or pos.y >= self.ymax) {
            return true;
        }
        return false;
    }

    fn dfs(self: *Self, pos: Pos, dirn: Direction, vis: *HashMap(Pos, Direction)) void {
        if (self.outside(pos)) return;

        const d = vis.get(pos);
        if (d != null and d.? == dirn) return;

        vis.put(pos, dirn) catch unreachable;
        const c = self.grid.items[@intCast(pos.y)].items[@intCast(pos.x)];
        switch (c) {
            .space => {
                switch (dirn) {
                    .up => self.dfs(pos.up(), dirn, vis),
                    .down => self.dfs(pos.down(), dirn, vis),
                    .left => self.dfs(pos.left(), dirn, vis),
                    .right => self.dfs(pos.right(), dirn, vis),
                }
            },
            .verticalSplitter => {
                switch (dirn) {
                    .up => self.dfs(pos.up(), dirn, vis),
                    .down => self.dfs(pos.down(), dirn, vis),
                    .left, .right => {
                        self.dfs(pos.up(), Direction.up, vis);
                        self.dfs(pos.down(), Direction.down, vis);
                    },
                }
            },
            .horizontalSplitter => {
                switch (dirn) {
                    .up, .down => {
                        self.dfs(pos.left(), Direction.left, vis);
                        self.dfs(pos.right(), Direction.right, vis);
                    },
                    .left => self.dfs(pos.left(), dirn, vis),
                    .right => self.dfs(pos.right(), dirn, vis),
                }
            },
            .leftMirror => {
                switch (dirn) {
                    .up => self.dfs(pos.left(), Direction.left, vis),
                    .down => self.dfs(pos.right(), Direction.right, vis),
                    .left => self.dfs(pos.up(), Direction.up, vis),
                    .right => self.dfs(pos.down(), Direction.down, vis),
                }
            },
            .rightMirror => {
                switch (dirn) {
                    .up => self.dfs(pos.right(), Direction.right, vis),
                    .down => self.dfs(pos.left(), Direction.left, vis),
                    .left => self.dfs(pos.down(), Direction.down, vis),
                    .right => self.dfs(pos.up(), Direction.up, vis),
                }
            },
        }
        return;
    }
};

fn part1(grid: *Grid) u32 {
    var hash = HashMap(Pos, Direction).init(grid.allocator);
    defer hash.deinit();
    grid.dfs(Pos.init(0, 0), Direction.right, &hash);
    return hash.count();
}

fn part2(grid: *Grid) u32 {
    var maxCount: u32 = 0;
    for (0..grid.xmax) |x| {
        var h = HashMap(Pos, Direction).init(grid.allocator);
        grid.dfs(Pos.init(@intCast(x), 0), Direction.down, &h);
        if (maxCount < h.count()) maxCount = h.count();
    }
    for (0..grid.xmax) |x| {
        var h = HashMap(Pos, Direction).init(grid.allocator);
        grid.dfs(Pos.init(@intCast(x), @intCast(grid.ymax - 1)), Direction.up, &h);
        if (maxCount < h.count()) maxCount = h.count();
    }
    for (0..grid.ymax) |y| {
        var h = HashMap(Pos, Direction).init(grid.allocator);
        grid.dfs(Pos.init(0, @intCast(y)), Direction.right, &h);
        if (maxCount < h.count()) maxCount = h.count();
    }
    for (0..grid.ymax) |y| {
        var h = HashMap(Pos, Direction).init(grid.allocator);
        grid.dfs(Pos.init(@intCast(grid.xmax - 1), @intCast(y)), Direction.left, &h);
        if (maxCount < h.count()) maxCount = h.count();
    }
    return maxCount;
}

pub fn main() !void {
    const s = time.microTimestamp();
    const stdin = std.io.getStdIn().reader();
    var grid = Grid.init(std.heap.page_allocator);
    defer grid.deinit();
    while (true) {
        var buf: [150]u8 = undefined;
        const line = try stdin.readUntilDelimiterOrEof(&buf, '\n');
        if (line == null) break;

        const l = line.?;
        var row = ArrayList(Cell).init(std.heap.page_allocator);
        for (0..l.len) |i| {
            const c = l[i];
            const cell = switch (c) {
                '.' => Cell.space,
                '|' => Cell.verticalSplitter,
                '-' => Cell.horizontalSplitter,
                '\\' => Cell.leftMirror,
                '/' => Cell.rightMirror,
                else => unreachable,
            };
            try row.append(cell);
        }
        try grid.append(row);
    }
    print("{d}\n", .{part1(&grid)});
    print("{d}\n", .{part2(&grid)});
    const e = time.microTimestamp();
    print("Time taken: {s}\n", .{std.fmt.fmtDurationSigned(e - s)});
}
